File: leetcode/algorithms/404_sum_of_left_leaves/main.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * Recursion (DFS)
 *
 * Complexities:
 *   N - The Numbder of Nodes in `root`
 *   H - The Height of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(H)
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumOfLeftLeaves = function (root) {
  if (!root) {
    return 0;
  }

  let result = 0;

  if (root.left) {
    if (!root.left.left && !root.left.right) {
      result += root.left.val;
    } else {
      result += sumOfLeftLeaves(root.left);
    }
  }

  result += sumOfLeftLeaves(root.right);

  return result;
};


// Solution
/**
 * Solution 1
 *
 * Recursion (DFS)
 *
 * Complexities:
 *   N - The Numbder of Nodes in `root`
 *   H - The Height of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(H)
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var solution1 = function (root) {
  if (!root) {
    return 0;
  }

  let sum = 0;
  const left = root.left;

  if (left && !left.left && !left.right) {
    sum += left.val;
  } else {
    sum += solution1(left);
  }

  sum += solution1(root.right);

  return sum;
};

/**
 * Solution 2
 *
 * Recursion (DFS) with Flag
 *
 * Complexities:
 *   N - The Numbder of Nodes in `root`
 *   H - The Height of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(H)
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var solution2 = function (root) {
  const dfs = (node, isLeft) => {
    if (!node) {
      return 0;
    }

    if (!node.left && !node.right) {
      return isLeft ? node.val : 0;
    }

    return dfs(node.left, true) + dfs(node.right, false);
  };

  return dfs(root, false);
};

/**
 * Solution 3
 *
 * Iteration (BFS) with Queue
 *
 * Complexities:
 *   N - The Numbder of Nodes in `root`
 *   W - The Width of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(W)
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var solution3 = function (root) {
  if (!root) {
    return 0;
  }

  let sum = 0;
  const queue = [root];
  let head = 0;

  while (head < queue.length) {
    const node = queue[head++];

    if (node.left) {
      if (!node.left.left && !node.left.right) {
        sum += node.left.val;
      } else {
        queue.push(node.left);
      }
    }

    if (node.right) {
      queue.push(node.right);
    }
  }

  return sum;
};

File: leetcode/algorithms/404_sum_of_left_leaves/main.java
import java.util.ArrayDeque;
import java.util.Deque;

class SumOfLeftLeaves {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    /**
     * Recursion (DFS)
     *
     * Complexities:
     *   N - The Numbder of Nodes in `root`
     *   H - The Height of `root`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(H)
     */
    public int sumOfLeftLeaves(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int result = 0;

        if (root.left != null) {
            if (root.left.left == null && root.left.right == null) {
                result += root.left.val;
            } else {
                result += sumOfLeftLeaves(root.left);
            }
        }

        result += sumOfLeftLeaves(root.right);

        return result;
    }


    // Solution
    /**
     * Solution 1
     * 
     * Recursion (DFS)
     *
     * Complexities:
     *   N - The Numbder of Nodes in `root`
     *   H - The Height of `root`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(H)
     */
    public int solution1(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int sum = 0;
        TreeNode left = root.left;

        if (left != null && left.left == null && left.right == null) {
            sum += left.val;
        } else {
            sum += sumOfLeftLeaves(left);
        }

        sum += sumOfLeftLeaves(root.right);

        return sum;
    }

    /**
     * Solution 2
     *
     * Iteration (BFS) with Deque
     *
     * Complexities:
     *   N - The Numbder of Nodes in `root`
     *   W - The Width of `root`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(W)
     */
    public int solution2(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int sum = 0;
        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();

            if (node.left != null) {
                if (node.left.left == null && node.left.right == null) {
                    sum += node.left.val;
                } else {
                    queue.offer(node.left);
                }
            }

            if (node.right != null) {
                queue.offer(node.right);
            }
        }

        return sum;
    }
}
